File: prototipo.cpp
#include <algorithm>
#include <cassert>
#include <chrono>
#include <fstream>
#include <iostream>
#include <memory>
#include <sstream>
#include <stack>
#include <string>
#include <utility>
#include <vector>

using std::size_t;

using time_point = std::chrono::high_resolution_clock::time_point;
using NodeId_t = size_t;
using real = double;

struct Node {
  NodeId_t num;
  size_t level;

  Node(size_t num, size_t level) : num(num), level(level) {}
};

struct Fuse {
  real R;
  real Imax;

  Fuse(real R, real Imax) : R(R), Imax(Imax) {}

  bool blown(real I) { return I > Imax; }
};

struct Edge {
  NodeId_t src;
  NodeId_t dst;
  Fuse fuse;

  Edge(NodeId_t src, NodeId_t dst, Fuse fuse) : src(src), dst(dst), fuse(fuse) {}
};

bool operator==(const Edge &a, const Edge &b) {
  return a.src == b.src && a.dst == b.dst;
}

template <typename C, typename T> auto erase(C &col, const T &val) {
  using std::begin;
  using std::end;
  return col.erase(std::remove(begin(col), end(col), val), end(col));
}

template <typename C, typename I>
auto swap_remove(C& container, I it) {
  using std::swap;
  swap(*it, container.back());
  container.pop_back();
  return it;
}

template <typename C>
auto swap_remove(C& container, size_t index) {
  using std::begin;
  return swap_remove(container, begin(container) + index);
}

struct Circuit {
  std::vector<NodeId_t> nodes;
  std::vector<std::vector<NodeId_t>> levels;
  std::vector<std::vector<Edge>> adjacencies;
  std::vector<std::vector<Edge>> inputs;

  Circuit(size_t n) : nodes(n), levels(), adjacencies(n), inputs(n) {}

  // void add_node(std::unique_ptr<Node> node) {
  void add_node(NodeId_t node, size_t level) {
    add_node_to_level(node, level);
    nodes[node] = node;
  }

  void add_node_to_level(NodeId_t node, size_t level) {
    while (levels.size() <= level) {
      levels.push_back({});
    }
    assert(levels.size() >= level + 1);
    levels[level].push_back(node);
  }

  void add_edge(const Edge &edge) {
    adjacencies[edge.src].push_back(edge);
    inputs[edge.dst].push_back(edge);
  }

  void remove_input_edge(const Edge& edge) {
    // erase(inputs[edge.dst], edge);
    auto& ins = inputs[edge.dst];
    swap_remove(ins, std::find(begin(ins), end(ins), edge));
  }

  bool connected() {
    auto source = nodes.front();
    auto sink = nodes.back();
    std::vector<char> visited(nodes.size(), 0);
    return idfs(source, sink, visited);
  }

  bool idfs(NodeId_t cur, NodeId_t target, std::vector<char> &visited) {
    std::stack<NodeId_t> s;
    s.push(cur);
    while (!s.empty()) {
      cur = s.top();
      s.pop();
      if (cur == target) return true;
      if (!visited[cur]) {
        visited[cur] = 1;
        for (auto& next : adjacencies[cur])
          s.push(next.dst);
      }
    }
    return false;
  }
};

const real inf = 1e9;

std::vector<real> calculate_parallel_current(const Circuit &g, real V) {
  auto R = 0.0;

  for (auto &adj : g.adjacencies) {
    for (auto &e : adj) {
      if (e.fuse.R != 0.0)
        R += 1.0 / e.fuse.R;
    }
  }

  auto I = V * R;
  return std::vector<real>(g.nodes.size(), I);
}

std::vector<real> calculate_polygon_current(const Circuit&g, real V) {
  std::vector<real> currents(g.nodes.size());
  std::vector<real> outputs(g.nodes.size());

  currents[0] = inf;
  outputs[0] = inf;

  for (auto i = 1u; i < g.levels.size(); i++) {
    for (auto& p : g.levels[i]) {
      for (auto& in : g.inputs[p]) {
        currents[p] += outputs[in.src]; // Resistencia impacta como?
      }
      auto num_outs = g.adjacencies[p].size();
      if (num_outs != 0)
        outputs[p] = currents[p] / num_outs; // Corrente distribuída igualmente entre as saídas
    }
  }

  return currents;
}

std::vector<real> calculate_series_current(const Circuit &g, real V) {
  auto R = 0.0;

  for (auto &adj : g.adjacencies) {
    for (auto &e : adj) {
      R += e.fuse.R;
    }
  }

  auto I = V / R;
  return std::vector<real>(g.nodes.size(), I);
}

Circuit generate_parallel_circuit(size_t L, real) {
  auto g = Circuit(L + 2);
  auto init = 0;
  g.add_node(init, 0);

  auto finish = L + 1;

  for (auto i = 1u; i < L; i += 2) {
    auto node_a = i;
    g.add_node(node_a, 1);
    g.add_edge(Edge(init, node_a, Fuse(0, inf)));

    auto node_b = i + 1;
    g.add_node(node_b, 2);
    g.add_edge(Edge(node_a, node_b, Fuse(20, 20 * i)));

    g.add_edge(Edge(node_b, finish, Fuse(0, inf)));
  }

  g.add_node(finish, 3);

  return g;
}

Circuit generate_series_circuit(size_t L, real) {
  auto g = Circuit(L + 2);

  auto init = 0;
  auto prev = init;
  g.add_node(init, 0);

  for (auto i = 1u; i <= L; i++) {
    auto node = i;
    g.add_node(node, i);

    auto R = i == 1 ? 0.0 : 20.0;
    auto Imax = i == 1 ? inf : 20.0;
    g.add_edge( Edge(prev, node, Fuse(R, Imax)));

    prev = node;
  }

  auto finish = L + 1;
  g.add_node(finish, L + 1);
  g.add_edge(Edge(prev, finish, Fuse(0, inf)));

  return g;
}

Circuit generate_square_circuit(size_t L, real D) {
  return Circuit(L);
}

Circuit generate_hexagon_circuit(size_t L, real D) {
  return Circuit(L);
}

enum TipoCircuito { Series, Parallel, Square, Hexagon };
TipoCircuito CurrentType = Parallel;

std::vector<real> calculate_current(const Circuit &g, real V) {
  switch (CurrentType) {
  case Parallel:
    return calculate_parallel_current(g, V);
  case Series:
    return calculate_series_current(g, V);
  case Square:
  case Hexagon:
    return calculate_polygon_current(g, V);
  default:
    return {};
  }
}

Circuit generate_circuit(size_t L, real D) {
  switch (CurrentType) {
  case Parallel:
    return generate_parallel_circuit(L, D);
  case Series:
    return generate_series_circuit(L, D);
  case Square:
    return generate_square_circuit(L, D);
  case Hexagon:
    return generate_hexagon_circuit(L, D);
  default:
    return Circuit(L);
  }
}

struct Log {
  std::vector<std::pair<real, real>> xy;

  void log(real V, real I) { xy.push_back({V, I}); }

  void show(std::ostream &out = std::cout) {
    out << "V,I\n";
    for (auto &p : xy) {
      out << p.first << "," << p.second << "\n";
    }
  }
};

void iteracao(Circuit &g, const std::vector<real>& currents) {
  for (auto &points : g.levels) {
    for (auto &p : points) {
      auto I = currents[p];
      auto &adj = g.adjacencies[p];
      auto it = begin(adj);

      while (it != end(adj)) {
        if (it->fuse.blown(I)) {
          g.remove_input_edge(*it);
          swap_remove(adj, it);
          // std::swap(*it, adj.back());
          // adj.pop_back();
        } else {
          ++it;
        }
      }
    }
  }
}

std::string flt2str(real f) {
  std::ostringstream oss;
  oss << std::defaultfloat << f;
  return oss.str();
}

void draw_graph(const Circuit &g) {
  std::ofstream out{"graph.dot"};
  out << "digraph {\n";
  out << "  graph [rankdir=LR]\n";
  out << "  node [height=0.05 label=\"\" shape=point width=0.05]\n";
  out << "  edge [arrowhead=empty arrowsize=0.5 fontsize=8]\n";

  for (auto &v : g.nodes) {
    for (auto &e : g.adjacencies[v]) {
      auto &f = e.fuse;
      std::string label;
      if (f.R == 0)
        label = "";
      else
        label = flt2str(f.R) + "Ω " + flt2str(f.Imax) + "A";
      out << "    " << v << " -> " << e.dst << "[label=\"" << label
          << "\"]\n";
    }
  }

  out << "}\n";
}

Log simulacao(size_t L, real D, real V0, real deltaV) {
  auto g = generate_circuit(L, D);
  // draw_graph(g);
  auto V = V0;
  auto l = Log();
  auto sink = g.nodes.size() - 1;

  while (g.connected()) {
    auto currents = calculate_current(g, V);
    l.log(V, currents[sink]);

    iteracao(g, currents);
    V += deltaV;
  }

  l.log(V, 0);
  return l;
}

time_point now() { return std::chrono::high_resolution_clock::now(); }

long elapsed(const time_point &t) {
  auto e = now() - t;
  return std::chrono::duration_cast<std::chrono::milliseconds>(e).count();
}

int main() {
  auto start = now();
  auto s = simulacao(400, 0, 0, 0.1);
  std::cout << elapsed(start) << " ms\n";
  std::ofstream out{"saida.csv"};
  s.show(out);
}

File: prototipo.py
"""
  fracture: Simulation of a scalar fracture model based on Freitas (2007).
  Copyright (C) 2017 Italo Silva

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

import matplotlib.pyplot as plt
import itertools
from graphviz import Digraph
#
import timeit
import cProfile
import pstats

inf = 1e9

def gerar_grid_serial(L, D):
    g = Circuito()

    init = Node(0, 0)
    g.add_node(init)

    prev = init
    for i in range(1, L + 1):
        node = Node(i, i)
        g.add_node(node)
        R = 0 if prev is init else 20
        Imax = inf if prev is init else 20
        g.add_edge(Edge(prev, node, Fusivel(R, Imax)))
        prev = node

    finish = Node(L + 1, L + 1)
    g.add_node(finish)
    g.add_edge(Edge(prev, finish, Fusivel(0, inf)))

    return g


def gerar_grid_paralelo(L, D):
    g = Circuito()

    init = Node(0, 0)
    g.add_node(init)

    finish = Node(L + 1, 3)

    for i in range(1, L, 2):
        node_a = Node(i, 1)
        g.add_node(node_a)
        g.add_edge(Edge(init, node_a, Fusivel(0, inf)))

        node_b = Node(i + 1, 2)
        g.add_node(node_b)
        g.add_edge(Edge(node_a, node_b, Fusivel(20, 20 * i)))

        g.add_edge(Edge(node_b, finish, Fusivel(0, inf)))

    g.add_node(finish)

    return g


def desenhar_grafo(g):
    dot = Digraph()
    dot.attr('graph', rankdir='LR')
    dot.attr('node', shape='point', label='', width='0.05', height='0.05')
    dot.attr('edge', arrowhead='empty', arrowsize='0.5', fontsize='8')

    for node in g.nodes:
        dot.node(str(node.num))

        for edge in g.adjacencias[node.num]:
            f = edge.fusivel
            if f.R == 0:
                label = ''
            else:
                label = f'{f.R}Ω, {f.Imax}A'
            dot.edge(str(node.num), str(edge.dst.num), label)

    print(dot.source)
    dot.render('test', view=True)


def gerar_grid(L, D):
    # return gerar_grid_serial(L, D)
    return gerar_grid_paralelo(L,D)


def calcula_corrente(g, V):
    # return calcula_corrente_serial(g, V)
    return calcula_corrente_paralelo(g, V)


def flatten(listOfLists):
    "Flatten one level of nesting"
    return itertools.chain.from_iterable(listOfLists)


def calcula_corrente_serial(g, V):
    edges = flatten(w for _, w in g.adjacencias.items())
    weights = (e.fusivel.R for e in edges)
    resistance = sum(weights)
    return V / resistance


def calcula_corrente_paralelo(g, V):
    edges = flatten(w for _, w in g.adjacencias.items())
    weights = (1 / e.fusivel.R for e in edges if e.fusivel.R != 0)
    resistance = sum(weights)
    return V * resistance


class Fusivel:
    def __init__(self, R, Imax):
        self.R = R
        self.Imax = Imax

    def queimou(self, I):
        return I > self.Imax


class Edge:
    def __init__(self, src, dst, fusivel):
        self.src = src
        self.dst = dst
        self.fusivel = fusivel


class Node:
    def __init__(self, num, nivel):
        self.num = num
        self.nivel = nivel


class Circuito:
    def __init__(self):
        self.nodes = []
        self.niveis = []
        self.adjacencias = {}

    def add_node(self, node):
        self.nodes.append(node)
        self.adjacencias[node.num] = []
        self.add_node_to_nivel(node)

    def add_node_to_nivel(self, node):
        if len(self.niveis) <= node.nivel:
            self.niveis.append([])
        assert (len(self.niveis) >= node.nivel + 1)
        self.niveis[node.nivel].append(node)

    def add_edge(self, edge):
        self.adjacencias[edge.src.num].append(edge)

    def remove(self, p):
        self.nodes.remove(p)
        self.niveis[p.nivel].remove(p)
        del self.adjacencias[p.num]

        for adj in self.adjacencias.values():
            for i, e in enumerate(adj):
                if e.dst == p:
                    del adj[i]
                    break

    def remove_edge(self, e):
        self.adjacencias[e.src.num].remove(e)

    def conectado(self):
        inicio = self.nodes[0]
        fim = self.nodes[-1]
        return self.dfs(inicio, fim, set())


    def dfs(self, cur, target, percorridos):
        if cur == target:
            return True

        percorridos.add(cur)
        for nxt in self.adjacencias[cur.num]:
            if nxt.dst not in percorridos and self.dfs(nxt.dst, target, percorridos):
                return True

        return False


def iteracao(g, V, deltaV, registro):
    """Tenta fazer uma iteração. Se alguém queimar, retorna a mesma
    ddp para que a simulação dessa rodada seja reiniciada com os
    fusíveis queimados. Se ninguém queimar, registra o resultado e
    retorna a próxima voltagem, para que comece a próxima etapa
    (uma etapa só termina quando a ddp muda, várias iterações podem
    acontecer na mesma etapa).
    """
    I = calcula_corrente(g, V)

    for pontos in g.niveis:
        queimado = False

        # A verificação é feita por níveis, de forma que fusíveis
        # nos mesmos níveis queimam juntos.
        for p in pontos:
            for e in g.adjacencias[p.num][:]:
                if e.fusivel.queimou(I):
                    # print(f'Queimou: {p.num}')
                    g.remove_edge(e)
                    queimado = True
        if queimado:
            return V

    registro.registrar(V, I)
    return V + deltaV


class Registro:
    def __init__(self):
        self.V = []
        self.I = []

    def registrar(self, V, I):
        self.V.append(V)
        self.I.append(I)
        # print(f'V: {V}, I: {I}')

    def plot(self):
        plt.plot(self.V, self.I)
        plt.show()


def simulacao(L, D, V0, deltaV):
    g = gerar_grid(L, D)
    desenhar_grafo(g)
    V = V0
    registro = Registro()

    while g.conectado():
        V = iteracao(g, V, deltaV, registro)

    registro.registrar(V, 0)
    return registro


def main():
    r = simulacao(20, 0, 0, 0.5)
    r.plot()

def profile():
    cProfile.run('simulacao(20, 0, 0, 0.1)', 'restats')
    p = pstats.Stats('restats')
    p.strip_dirs().sort_stats('time').print_stats()


def time():
    n = 1
    t = timeit.timeit('simulacao(20, 0, 0, 0.1)', globals=globals(), number=n)
    print(f'Total: {t}s, média: {t/n}s')


if __name__ == '__main__':
    # main()
    profile()
    # time()
